src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1277,28 +1252,48 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
 
         // Only render attributes if client-side validation is enabled, and then only if we've
         // never rendered validation for a field with this name in this form.
-        protected virtual IDictionary<string, object> GetValidationAttributes(
+        public virtual void AddValidationAttributes(
```

Comment:
Let's go w/ `protected virtual` unless that breaks lots of tests. In that case, leave as you've got it.

Please add doc comments.